banner/models.py

```diff
@@ -1,189 +1,37 @@
-    )
-    width = models.PositiveIntegerField()
-    height = models.PositiveIntegerField()
-    targeting_key = models.CharField(
-        max_length=128,
-        help_text="Eg. 'interests'."
+class Banner(ModelBase):
+    """Base class for all banners"""
+    link = models.ForeignKey(
+        Link, help_text=_("Link to which this banner should redirect."),
+        blank=True, null=True
     )
-    targeting_values = models.TextField(
-        help_text="""Eg. 'sports'. One entry per line. Entries are allowed to \
-contain spaces."""
-        )
-
-    class Meta:
-        abstract = True
-
-    @property
-    def targeting_values_list(self):
-        return self.targeting_values.split('\r\n')
-
-
-class DFPBanner(BaseDFPBanner):
-    pass
+    style = models.CharField(choices=[(klass.__name__, klass.__name__) for klass in BANNER_STYLE_CLASSES], max_length=128)
 
 
-class BannerProxy(ModelBase):
-    """A banner that inspects the path and renders another banner"""
-    banners = models.ManyToManyField(
-        Banner,
-        blank=True,
-        help_text="""A list of banners which are proxied by this item. It is \
-useful if a page contains more than one banner proxy."""
+class Button(models.Model):
+    """Call to action handling"""
+    text = models.CharField(
+        max_length=60,
+        help_text=_("The text to be displayed as the button label")
     )
-    default_banner = models.ForeignKey(
-        Banner,
-        blank=True,
-        null=True,
-        related_name='bannerproxy_default_banner'
+    link = models.ForeignKey(
+        Link, help_text=_("CTA link for this button"), null=True, blank=True
     )
+    banner = models.ManyToManyField(to=Banner, related_name="buttons", null=True, blank=True, through="ButtonOrder")
 
-    class Meta:
-        verbose_name_plural = 'Banner proxies'
-
-    def get_actual_banner(self, request):
-        """Return first banner matching the path"""
-
-        # Try our set of banners
-        request_path = request.get_full_path()
-        path_banner_pairs = []
-        banners = Banner.permitted.filter(
-            id__in=self.banners.all()
-        ).order_by('?')
-        for banner in banners:
-            if banner.paths:
-                for path in banner.paths.split():
-                    path_banner_pairs.append((path, banner))
-
-        # Sort banner pairs because we want the best regex match
-        path_banner_pairs.sort(lambda a, b: cmp(len(b[0]), len(a[0])))
-
-        # Return first match
-        for path, banner in path_banner_pairs:
-            if re.search(r'%s' % path, request_path):
-                return banner.as_leaf_class()
 
-        # Fall back to default banner
-        if self.default_banner:
-            try:
-                default_banner = Banner.permitted.get(id=self.default_banner.id)
-                return default_banner.as_leaf_class()
-            except Banner.DoesNotExist:
-                pass
+class ButtonOrder(models.Model):
```

Comment:
I found we use `sortedm2m.fields.SortedManyToManyField` in jmbo itself, might be worth looking into.